src/tests/test-jsonrpc-fork-lua

```diff
@@ -0,0 +1,172 @@
+    end },
+    { "^%-%-remote%-address%=(.*)$", function(o)
+        if not o or #o < 1 then return false end
+        remote_address = o
+        return true
+    end },
+    { ".*", function(all)
+        error("unrecognized option " .. all)
+    end }
+}
+
+-- Process command line options
+local arguments = {}
+for i, argument in ipairs({...}) do
+    if argument:sub(1,1) == "-" then
+        for j, option in ipairs(options) do
+            if option[2](argument:match(option[1])) then
+                break
+            end
+        end
+    else
+        arguments[#arguments+1] = argument
+    end
+end
+
+local remote = jsonrpc.stub(remote_address)
+
+-- This is a simple test that creates a tree of forked servers
+-- Each new child of a given server modifies an additional register
+
+-- The root will have x1--x31 set to zero initially
+-- It will then modify x1 to 1 and fork
+-- Then x2 to 1 and fork
+-- Then x3 to 1 and fork
+-- So the ith child will have registers x1-xi set to 1
+-- Each child will then do the same, but child i will
+
+local FANOUT = 3
+local MAXDEPTH = 3
+
+local function new_x()
+    local t = {}
+    for i = 1, 31 do
+        t[i] = 0
+    end
+    return t
+end
+
+local function clone_x(x)
+    local y = {}
+    for i = 1, 31 do
+        y[i] = x[i]
+    end
+    return y
+end
+
+local function fork_tree(address, x, depth)
+    local stub = assert(jsonrpc.stub(address))
+    local node = {
+        address = address,
+        stub = stub,
+        x = x
+    }
+    local machine = stub.get_machine()
+    local children = {}
+    node.children = children
+    if depth <= MAXDEPTH then
+        for child_index = 1, FANOUT do
+            machine:write_x(child_index, depth)
+            x[child_index] = depth
+            local child = fork_tree(stub.fork(), clone_x(x), depth + 1)
+            children[#children+1] = child
+        end
+    end
+    return node
+end
+
+local function pre_order(node, f, depth)
+    depth = depth or 0
+    f(node, depth)
+    for i, child in ipairs(node.children) do
+        pre_order(child, f, depth+1)
+    end
+end
+
+local function check_tree(root)
+    pre_order(root, function(node, depth)
+        local machine = node.stub.get_machine()
+        local x = node.x
+        io.write(string.rep("  ", depth), "{", table.concat(node.x, ","), "}\n")
+        for i = 1, 31 do
+            if machine:read_x(i) ~= x[i] then
+                error("mismatch in x[" .. i .. "]")
+            end
+        end
+    end)
+end
+
+local function kill_tree(root)
+    pre_order(root, function(node)
+        node.stub.shutdown()
+    end)
+end
+
+local function tmprom()
+    local name = os.tmpname()
+    local f = io.open(name, "wb")
+    f:write(string.rep("\0", 4096))
+    f:close()
+    return name
+end
+
+local stub = jsonrpc.stub(remote_address)
+local config = stub.machine.get_default_config()
+config.rom.image_filename = tmprom()
+config.ram.length = 1 << 22
+
+local machine = stub.machine(config)
```

Comment:
This variable is set but unused (caught by luacheck). But of course removing this line will break the test.

Feels little odd to instantiate a machine and not use it, but I noticed the machine is retrieved later with `stub.get_machine()`.

Reply:
If I don't initialize the machine here, I can initialize it in the recursive function. It's a bit more awkward, but can do.